tempesta_fw/t/unit/tfw_fuzzer.c

```diff
@@ -0,0 +1,563 @@
+	} else {
+		if (gen_vector[CONTENT_LENGTH].i % 2) {
+			add_string(s1, gen_vector[METHOD].i,
+				   gen_vector[METHOD].a_val);
+			return FUZZ_VALID;
+		} else
+			add_string(s1, gen_vector[METHOD].i,
+				   gen_vector[METHOD].a_inval);
+	}
+
+	return FUZZ_INVALID;
+}
+
+static int add_uri_path_start(char *s1)
+{
+	struct fuzz_msg r = uri_path_start[gen_vector[URI_PATH_START].i];
+
+	if (r.s) {
+		strcat(s1, r.s);
+		return r.inval;
+	} else {
+		if (gen_vector[CONTENT_LENGTH].i % 2) {
+			add_string(s1, gen_vector[URI_PATH_START].i,
+				   gen_vector[URI_PATH_START].a_val);
+			return FUZZ_VALID;
+		} else
+			add_string(s1, gen_vector[URI_PATH_START].i,
+				   gen_vector[URI_PATH_START].a_inval);
+	}
+
+	return 1;
+}
+
+static int add_uri_file(char *s1)
+{
+	struct fuzz_msg r = uri_file[gen_vector[URI_FILE].i];
+
+	if (r.s) {
+		strcat(s1, r.s);
+		return r.inval;
+	} else {
+		if (gen_vector[CONTENT_LENGTH].i % 2) {
+			add_string(s1, gen_vector[URI_FILE].i,
+				   gen_vector[URI_FILE].a_val);
+			return FUZZ_VALID;
+		} else
+			add_string(s1, gen_vector[URI_FILE].i,
+				   gen_vector[URI_FILE].a_inval);
+	}
+
+	return FUZZ_INVALID;
+}
+
+static int add_http_ver(char *s1)
+{
+	struct fuzz_msg r = versions[gen_vector[HTTP_VER].i];
+
+	if (r.s) {
+		strcat(s1, r.s);
+		return r.inval;
+	} else {
+		if (gen_vector[CONTENT_LENGTH].i % 2) {
+			add_string(s1, gen_vector[HTTP_VER].i,
+				   gen_vector[HTTP_VER].a_val);
+			return FUZZ_VALID;
+		} else
+			add_string(s1, gen_vector[HTTP_VER].i,
+				   gen_vector[HTTP_VER].a_inval);
+	}
+
+	return FUZZ_INVALID;
+}
+
+static int add_resp_code(char *s1)
+{
+	struct fuzz_msg r = resp_code[gen_vector[RESP_CODE].i];
+
+	if (r.s) {
+		strcat(s1, r.s);
+		return r.inval;
+	} else {
+		if (gen_vector[CONTENT_LENGTH].i % 2) {
+			add_string(s1, gen_vector[RESP_CODE].i,
+				   gen_vector[RESP_CODE].a_val);
+			return FUZZ_VALID;
+		} else
+			add_string(s1, gen_vector[RESP_CODE].i,
+				   gen_vector[RESP_CODE].a_inval);
+	}
+
+	return FUZZ_INVALID;
+}
+
+static int add_connection(char *s1)
+{
+	struct fuzz_msg r = conn_val[gen_vector[CONNECTION].i];
+
+	strcat(s1, "Connection:");
+	add_spaces(s1);
+	if (r.s) {
+		strcat(s1, r.s);
+		return r.inval;
+	} else {
+		if (gen_vector[CONTENT_LENGTH].i % 2) {
+			add_string(s1, gen_vector[CONNECTION].i,
+				   gen_vector[CONNECTION].a_val);
+			return FUZZ_VALID;
+		} else
+			add_string(s1, gen_vector[CONNECTION].i,
+				   gen_vector[CONNECTION].a_inval);
+	}
+
+	return FUZZ_INVALID;
+}
+
+static int add_user_agent(char *s1)
+{
+	struct fuzz_msg r = ua_val[gen_vector[USER_AGENT].i];
+
+	strcat(s1, "User-agent:");
+	add_spaces(s1);
+	if (r.s) {
+		strcat(s1, r.s);
+		return r.inval;
+	} else {
+		if (gen_vector[CONTENT_LENGTH].i % 2) {
+			add_string(s1, gen_vector[USER_AGENT].i,
+				   gen_vector[USER_AGENT].a_val);
+			return FUZZ_VALID;
+		} else
+			add_string(s1, gen_vector[USER_AGENT].i,
+				   gen_vector[USER_AGENT].a_inval);
+	}
```

Comment:
Some time ago we saw very large User-Agent value, it was Internet Explorer with sequence of tons of updates. It hase several kilobytes in size. Generally speaking, it's good to generate such large values for all headers at least once.